# Classifying the classes of wine from the wine dataset

# Data preprocessing 

In [89]:
import pandas as pd
#importing the dataset into pandas dataframe 
df = pd.read_csv ("wine.csv")
df

1  14.23  1.71  2.43  15.6  127   2.8  3.06  0.28  2.29       5.64  1.04  \
0    1  13.20  1.78  2.14  11.2  100  2.65  2.76  0.26  1.28   4.380000  1.05   
1    1  13.16  2.36  2.67  18.6  101  2.80  3.24  0.30  2.81   5.680000  1.03   
2    1  14.37  1.95  2.50  16.8  113  3.85  3.49  0.24  2.18   7.800000  0.86   
3    1  13.24  2.59  2.87  21.0  118  2.80  2.69  0.39  1.82   4.320000  1.04   
4    1  14.20  1.76  2.45  15.2  112  3.27  3.39  0.34  1.97   6.750000  1.05   
5    1  14.39  1.87  2.45  14.6   96  2.50  2.52  0.30  1.98   5.250000  1.02   
6    1  14.06  2.15  2.61  17.6  121  2.60  2.51  0.31  1.25   5.050000  1.06   
7    1  14.83  1.64  2.17  14.0   97  2.80  2.98  0.29  1.98   5.200000  1.08   
8    1  13.86  1.35  2.27  16.0   98  2.98  3.15  0.22  1.85   7.220000  1.01   
9    1  14.10  2.16  2.30  18.0  105  2.95  3.32  0.22  2.38   5.750000  1.25   
10   1  14.12  1.48  2.32  16.8   95  2.20  2.43  0.26  1.57   5.000000  1.17   
11   1  13.75  1.73  2.41  16.0   89  2.60  2.76  0.29  1.81   5.600000  1.15   
12   1  14.75  1.73  2.39  11.4   91  3.10  3.69  0.43  2.81   5.400000  1.25   
13   1  14.38  1.87  2.38  12.0  102  3.30  3.64  0.29  2.96   7.500000  1.20   
14   1  13.63  1.81  2.70  17.2  112  2.85  2.91  0.30  1.46   7.300000  1.28   
15   1  14.30  1.92  2.72  20.0  120  2.80  3.14  0.33  1.97   6.200000  1.07   
16   1  13.83  1.57  2.62  20.0  115  2.95  3.40  0.40  1.72   6.600000  1.13   
17   1  14.19  1.59  2.48  16.5  108  3.30  3.93  0.32  1.86   8.700000  1.23   
18   1  13.64  3.10  2.56  15.2  116  2.70  3.03  0.17  1.66   5.100000  0.96   
19   1  14.06  1.63  2.28  16.0  126  3.00  3.17  0.24  2.10   5.650000  1.09   
20   1  12.93  3.80  2.65  18.6  102  2.41  2.41  0.25  1.98   4.500000  1.03   
21   1  13.71  1.86  2.36  16.6  101  2.61  2.88  0.27  1.69   3.800000  1.11   
22   1  12.85  1.60  2.52  17.8   95  2.48  2.37  0.26  1.46   3.930000  1.09   
23   1  13.50  1.81  2.61  20.0   96  2.53  2.61  0.28  1.66   3.520000  1.12   
24   1  13.05  2.05  3.22  25.0  124  2.63  2.68  0.47  1.92   3.580000  1.13   
25   1  13.39  1.77  2.62  16.1   93  2.85  2.94  0.34  1.45   4.800000  0.92   
26   1  13.30  1.72  2.14  17.0   94  2.40  2.19  0.27  1.35   3.950000  1.02   
27   1  13.87  1.90  2.80  19.4  107  2.95  2.97  0.37  1.76   4.500000  1.25   
28   1  14.02  1.68  2.21  16.0   96  2.65  2.33  0.26  1.98   4.700000  1.04   
29   1  13.73  1.50  2.70  22.5  101  3.00  3.25  0.29  2.38   5.700000  1.19   
..  ..    ...   ...   ...   ...  ...   ...   ...   ...   ...        ...   ...   
147  3  13.32  3.24  2.38  21.5   92  1.93  0.76  0.45  1.25   8.420000  0.55   
148  3  13.08  3.90  2.36  21.5  113  1.41  1.39  0.34  1.14   9.400000  0.57   
149  3  13.50  3.12  2.62  24.0  123  1.40  1.57  0.22  1.25   8.600000  0.59   
150  3  12.79  2.67  2.48  22.0  112  1.48  1.36  0.24  1.26  10.800000  0.48   
151  3  13.11  1.90  2.75  25.5  116  2.20  1.28  0.26  1.56   7.100000  0.61   
152  3  13.23  3.30  2.28  18.5   98  1.80  0.83  0.61  1.87  10.520000  0.56   
153  3  12.58  1.29  2.10  20.0  103  1.48  0.58  0.53  1.40   7.600000  0.58   
154  3  13.17  5.19  2.32  22.0   93  1.74  0.63  0.61  1.55   7.900000  0.60   
155  3  13.84  4.12  2.38  19.5   89  1.80  0.83  0.48  1.56   9.010000  0.57   
156  3  12.45  3.03  2.64  27.0   97  1.90  0.58  0.63  1.14   7.500000  0.67   
157  3  14.34  1.68  2.70  25.0   98  2.80  1.31  0.53  2.70  13.000000  0.57   
158  3  13.48  1.67  2.64  22.5   89  2.60  1.10  0.52  2.29  11.750000  0.57   
159  3  12.36  3.83  2.38  21.0   88  2.30  0.92  0.50  1.04   7.650000  0.56   
160  3  13.69  3.26  2.54  20.0  107  1.83  0.56  0.50  0.80   5.880000  0.96   
161  3  12.85  3.27  2.58  22.0  106  1.65  0.60  0.60  0.96   5.580000  0.87   
162  3  12.96  3.45  2.35  18.5  106  1.39  0.70  0.40  0.94   5.280000  0.68   
163  3  13.78  2.76  2.30  22.0   90  1.35  0.68  0.41  1.03   9.580000  0.70   
164  3  13.73  4.36  2.26  22.5   88

In [94]:
#naming the columns in the data frame
df.columns=['class','Alcohol','Malicacid','Ash','Alcalinityofash','Magnesium','Totalphenols','Flavanoids','Nonflavanoidphenols','Proanthocyanins','Colorintensity','Hue','OD280/OD315','Proline']
df

class  Alcohol  Malicacid   Ash  Alcalinityofash  Magnesium  \
0        1    13.20       1.78  2.14             11.2        100   
1        1    13.16       2.36  2.67             18.6        101   
2        1    14.37       1.95  2.50             16.8        113   
3        1    13.24       2.59  2.87             21.0        118   
4        1    14.20       1.76  2.45             15.2        112   
5        1    14.39       1.87  2.45             14.6         96   
6        1    14.06       2.15  2.61             17.6        121   
7        1    14.83       1.64  2.17             14.0         97   
8        1    13.86       1.35  2.27             16.0         98   
9        1    14.10       2.16  2.30             18.0        105   
10       1    14.12       1.48  2.32             16.8         95   
11       1    13.75       1.73  2.41             16.0         89   
12       1    14.75       1.73  2.39             11.4         91   
13       1    14.38       1.87  2.38             12.0        102   
14       1    13.63       1.81  2.70             17.2        112   
15       1    14.30       1.92  2.72             20.0        120   
16       1    13.83       1.57  2.62             20.0        115   
17       1    14.19       1.59  2.48             16.5        108   
18       1    13.64       3.10  2.56             15.2        116   
19       1    14.06       1.63  2.28             16.0        126   
20       1    12.93       3.80  2.65             18.6        102   
21       1    13.71       1.86  2.36             16.6        101   
22       1    12.85       1.60  2.52             17.8         95   
23       1    13.50       1.81  2.61             20.0         96   
24       1    13.05       2.05  3.22             25.0        124   
25       1    13.39       1.77  2.62             16.1         93   
26       1    13.30       1.72  2.14             17.0         94   
27       1    13.87       1.90  2.80             19.4        107   
28       1    14.02       1.68  2.21             16.0         96   
29       1    13.73       1.50  2.70             22.5        101   
..     ...      ...        ...   ...              ...        ...   
147      3    13.32       3.24  2.38             21.5         92   
148      3    13.08       3.90  2.36             21.5        113   
149      3    13.50       3.12  2.62             24.0        123   
150      3    12.79       2.67  2.48             22.0        112   
151      3    13.11       1.90  2.75             25.5        116   
152      3    13.23       3.30  2.28             18.5         98   
153      3    12.58       1.29  2.10             20.0        103   
154      3    13.17       5.19  2.32             22.0         93   
155      3    13.84       4.12  2.38             19.5         89   
156      3    12.45       3.03  2.64             27.0         97   
157      3    14.34       1.68  2.70             25.0         98   
158      3    13.48       1.67  2.64             22.5         89   
159      3    12.36       3.83  2.38             21.0         88   
160      3    13.69       3.26  2.54             20.0        107   
161      3    12.85       3.27  2.58             22.0        106   
162      3    12.96       3.45  2.35             18.5        106   
163      3    13.78       2.76  2.30             22.0         90   
164      3    13.73       4.36  2.26             22.5         88   
165      3    13.45       3.70  2.60             23.0        111   
166      3    12.82       3.37  2.30             19.5         88   
167      3    13.58       2.58  2.69             24.5        105   
168      3    13.40       4.60  2.86             25.0        112   
169      3    12.20       3.03  2.32             19.0         96   
170      3    12.77       2.39  2.28             19.5         86   
171      3    14.16       2.51  2.48             20.0         91   
172      3    13.71       5.65  2.45             20.5         95   
173      3    13.40       3.91  2.48             23.0        

In [95]:
#Finding mean,minimum and maximum values to find the ouliers
df.describe()

class     Alcohol   Malicacid         Ash  Alcalinityofash  \
count  177.000000  177.000000  177.000000  177.000000       177.000000   
mean     1.943503   12.993672    2.339887    2.366158        19.516949   
std      0.773991    0.808808    1.119314    0.275080         3.336071   
min      1.000000   11.030000    0.740000    1.360000        10.600000   
25%      1.000000   12.360000    1.600000    2.210000        17.200000   
50%      2.000000   13.050000    1.870000    2.360000        19.500000   
75%      3.000000   13.670000    3.100000    2.560000        21.500000   
max      3.000000   14.830000    5.800000    3.230000        30.000000   

        Magnesium  Totalphenols  Flavanoids  Nonflavanoidphenols  \
count  177.000000    177.000000  177.000000           177.000000   
mean    99.587571      2.292260    2.023446             0.362316   
std     14.174018      0.626465    0.998658             0.124653   
min     70.000000      0.980000    0.340000             0.130000   
25%     88.000000      1.740000    1.200000             0.270000   
50%     98.000000      2.350000    2.130000             0.340000   
75%    107.000000      2.800000    2.860000             0.440000   
max    162.000000      3.880000    5.080000             0.660000   

       Proanthocyanins  Colorintensity         Hue  OD280/OD315      Proline  
count       177.000000      177.000000  177.000000   177.000000   177.000000  
mean          1.586949        5.054802    0.956983     2.604294   745.096045  
std           0.571545        2.324446    0.229135     0.705103   314.884046  
min           0.410000        1.280000    0.480000     1.270000   278.000000  
25%           1.250000        3.210000    0.780000     1.930000   500.000000  
50%           1.550000        4.680000    0.960000     2.780000   672.000000  
75%           1.950000        6.200000    1.120000     3.170000   985.000000  
max           3.580000       13.000000    1.710000     4.000000  1680.000000

There are no outliers in the data as all the values in the respective columns are close to the mean value of the columns

# Scaling the data

In [99]:
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
scaler.fit(df)

MinMaxScaler(copy=True, feature_range=(0, 1))

Scaled the dataset using MinMaxScaler 

# Decision tree

In [100]:
from sklearn import tree
#converting the pandas dataframe to arrays
X = df[df.columns[1:13]]
y = df["class"]
clf = DecisionTreeClassifier()
clf = clf.fit(X,y)

In [101]:
from sklearn.model_selection import train_test_split  
from sklearn.metrics import accuracy_score 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 100) 

In [102]:
y_pred_DecisionTreeClassifier = clf.predict(X_test) 
print ("Desicion Tree Accuracy is ",accuracy_score(y_test, y_pred_DecisionTreeClassifier)*100 )

Desicion Tree Accuracy is  100.0


In [103]:
y_pred_DecisionTreeClassifier

array([1, 2, 1, 2, 3, 3, 2, 3, 1, 2, 1, 3, 3, 2, 2, 1, 3, 1, 1, 3, 2, 2,
       3, 1, 3, 3, 2, 3, 3, 3, 1, 2, 2, 3, 3, 3, 2, 2, 3, 1, 2, 3, 2, 2,
       2], dtype=int64)

In [104]:
#Using 10 fold cross validation
from sklearn.model_selection import cross_val_score
print(cross_val_score(clf, X, y, cv=10))
Decision_tree_scores=cross_val_score(clf, X, y, cv=10)

[0.94736842 0.83333333 0.88888889 0.72222222 0.88888889 1.
 0.94444444 0.94444444 1.         1.        ]


In [62]:
import statistics
statistics.mean(Decision_tree_scores)*100

90.57383040935673

Accuracy of the Decision tree while using 10 fold cross validation is 90.57383040935673

# Random Forest

In [105]:
from sklearn.ensemble import RandomForestClassifier

In [106]:
randomforest =RandomForestClassifier(n_estimators=10)

In [107]:
randomforest.fit(X,y)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
                       max_depth=None, max_features='auto', max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=10,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [108]:
y_pred_DecisionTreeClassifier = randomforest.predict(X_test) 
print ("Random Forest Accuracy is ",accuracy_score(y_test, y_pred_DecisionTreeClassifier)*100 )

Random Forest Accuracy is  100.0


In [111]:
from sklearn.model_selection import cross_val_score
print(cross_val_score(clf, X, y, cv=10))
Random_Forest_scores=cross_val_score(clf, X, y, cv=10)
statistics.mean(Random_Forest_scores)*100

[0.84210526 0.83333333 0.88888889 0.72222222 0.88888889 1.
 0.94444444 1.         1.         1.        ]


91.68494152046783

Accuracy of the Random Forest using 10 fold cross validation is 91.68494152046783.Accuracy of the Random Forest is higher than Decision tree

# KNN classifier

In [112]:
from sklearn.neighbors import KNeighborsClassifier

In [113]:
knnclassifier =KNeighborsClassifier(n_neighbors=10)

In [114]:
knn_classifier_scores=cross_val_score(knnclassifier, X, y, cv=10)
statistics.mean(knn_classifier_scores)

0.770796783625731

In [115]:
knnclassifier =KNeighborsClassifier()

In [117]:
for k in range(1,50):
    knnclassifier =KNeighborsClassifier(n_neighbors=k)
    knn_classifier_scores=cross_val_score(knnclassifier, X, y, cv=10)
    print(statistics.mean(knn_classifier_scores)*100)

85.97953216374269
80.96856725146199
87.02119883040935
83.19078947368422
83.12134502923976
79.78801169590643
80.31432748538012
80.96856725146199
78.1907894736842
77.0796783625731
77.7046783625731
77.63523391812865
76.52412280701755
77.10891812865496
77.10891812865496
77.66447368421052
77.03947368421052
75.95760233918128
74.1922514619883
74.29093567251462
74.36038011695906
72.06871345029239
73.1798245614035
73.73538011695906
73.1798245614035
71.4437134502924
71.4437134502924
70.33260233918128
69.22149122807018
69.77704678362574
68.04093567251462
68.04093567251462
65.74926900584795
65.72002923976609
65.72002923976609
65.74926900584795
65.74926900584795
65.72002923976609
65.16447368421052
66.34502923976609
64.05336257309942
62.942251461988306
63.49780701754386
63.49780701754386
64.12280701754386
63.49780701754386
62.3172514619883
62.942251461988306
62.942251461988306


Accuracy of KNN is changing with change in K.Value of K makes a substantial difference in KNN model.Accuracy of the KNN is high when k=3.Accuracy of KNN is 87.02119883040935

# Naive Bayes

# MultinomialNB

In [84]:
from sklearn.naive_bayes import MultinomialNB

In [85]:
MultinomialNB_classifier =MultinomialNB()
MultinomialNB_classifier_scores=cross_val_score(MultinomialNB_classifier, X, y, cv=10)
statistics.mean(MultinomialNB_classifier_scores)*100

86.87134502923976

Accuracy of MultinomialNB with 10 fold cross validation is 86.87134502923976

# GaussianNB

In [86]:
from sklearn.naive_bayes import GaussianNB

In [87]:
GaussianNB_classifier =GaussianNB()
GaussianNB_classifier_scores=cross_val_score(GaussianNB_classifier, X, y, cv=10)
statistics.mean(GaussianNB_classifier_scores)*100

94.50292397660819

Accuracy of GaussianNB with 10 fold cross validation is 94.50292397660819.Accuracy of GaussianNB is higher than MultinomialNB

Conclusion: All the models are classifying well. Among all the models GaussianNB is having the highest accuracy and multinomial NB is having the lowest accuracy.